server/src/main/java/org/elasticsearch/action/bulk/BulkProcessor2.java

```diff
@@ -0,0 +1,419 @@
+                        + " because "
+                        + totalBytesInFlight.get()
+                        + " bytes are already in flight and the max is "
+                        + maxBytesInFlight
+                );
+            }
+            long bytesBeforeNewRequest = bulkRequest.estimatedSizeInBytes();
+            bulkRequest.add(request);
+            totalBytesInFlight.addAndGet(bulkRequest.estimatedSizeInBytes() - bytesBeforeNewRequest);
+            bulkRequestToExecute = newBulkRequestIfNeeded();
+        }
+        // execute sending the local reference outside the lock to allow handler to control the concurrency via it's configuration.
+        if (bulkRequestToExecute != null) {
+            execute(bulkRequestToExecute.v1(), bulkRequestToExecute.v2());
+        }
+        scheduleFlushTask();
+    }
+
+    /**
+     * This method schedules a flush task to run flushInterval in the future if flushInterval is not null and if there is not already a
+     * flush task scheduled.
+     */
+    private void scheduleFlushTask() {
+        if (flushInterval == null) {
+            return;
+        }
+        /*
+         * This method is called from multiple threads. We synchronize on mutex here so that we are sure that cancellableFlushTask is not
+         * changed between when we check it and when we set it (whether that is a transition from null -> not null from another thread
+         * in this method or a change from not null -> null from the scheduled task).
+         */
+        synchronized (mutex) {
+            if (cancellableFlushTask == null) {
+                cancellableFlushTask = threadPool.schedule(() -> {
+                    synchronized (mutex) {
+                        if (closed == false && bulkRequest.numberOfActions() > 0) {
+                            execute();
+                        }
+                        cancellableFlushTask = null;
+                    }
+                }, flushInterval, ThreadPool.Names.GENERIC);
+            }
+        }
+    }
+
+    private Tuple<BulkRequest, Long> newBulkRequestIfNeeded() {
+        assert Thread.holdsLock(mutex);
+        ensureOpen();
+        if (isOverTheLimit() == false) {
+            return null;
+        }
+        final BulkRequest bulkRequest = this.bulkRequest;
+        this.bulkRequest = new BulkRequest();
+        return new Tuple<>(bulkRequest, executionIdGen.incrementAndGet());
+    }
+
+    /**
+     * This method sends the bulkRequest to the consumer up to maxNumberOfRetries times. The executionId is used to notify the listener
+     * both before and after the request.
+     * @param bulkRequest
+     * @param executionId
+     */
+    private void execute(BulkRequest bulkRequest, long executionId) {
+        // We're taking it off the books as far as BulkProcessor2 is concerned, and letting Retry2 account for it from here on out:
+        totalBytesInFlight.addAndGet(-1 * bulkRequest.estimatedSizeInBytes());
```

Comment:
Ah I intended to suggest tracking the bytes here _and only here_, releasing them alongside the calls to `listener.afterBulk` below. No need for any accounting in `Retry2` if we do that I think.

Reply:
Oh so if 99% of a bulk is complete and we're just retrying that last 1%, consider it all still in flight since we're still holding onto the responses? I guess that makes sense.